jenkins/bootstrap.py

```diff
@@ -0,0 +1,438 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Bootstraps starting a test job.
+
+The following should already be done:
+  git checkout http://k8s.io/test-infra
+  cd $WORKSPACE
+  test-infra/jenkins/bootstrap.py <--repo=R> <--job=J> <--pull=P || --branch=B>
+
+The bootstrapper now does the following:
+  # Note start time
+  # read test-infra/jenkins/$JOB.json
+  # check out repoes defined in $JOB.json
+  # note job started
+  # call runner defined in $JOB.json
+  # upload artifacts (this will change later)
+  # upload build-log.txt
+  # note job ended
+
+The contract with the runner is as follows:
+  * Runner must exit non-zero if job fails for any reason.
+"""
+
+
+import argparse
+import json
+import logging
+import os
+import select
+import socket
+import subprocess
+import sys
+import time
+
+
+ORIG_CWD = os.getcwd()  # Checkout changes cwd
+
+
+def Subprocess(cmd, stdin=None, check=True, output=None):
+    logging.info('Call subprocess:\n  %s', ' '.join(cmd))
+    proc = subprocess.Popen(
+        cmd,
+        stdin=subprocess.PIPE if stdin is not None else None,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    if stdin:
+        proc.stdin.write(stdin)
+        proc.stdin.close()
+    out = []
+    code = None
+    reads = [proc.stdout.fileno(), proc.stderr.fileno()]
+    while reads:
+        ret = select.select(reads, [], [], 0.1)
+        for fd in ret[0]:
+            if fd == proc.stdout.fileno():
+                line = proc.stdout.readline()
+                if not line:
+                    reads.remove(fd)
+                    continue
+                logging.info(line[:-1])
+                if output:
+                    out.append(line)
+            if fd == proc.stderr.fileno():
+                line = proc.stderr.readline()
+                if not line:
+                    reads.remove(fd)
+                    continue
+                logging.warning(line[:-1])
+
+    code = proc.wait()
+    logging.info('Subprocess %d exited with code %d' % (proc.pid, code))
+    lines = output and '\n'.join(out)
+    if check and code:
+        raise subprocess.CalledProcessError(code, cmd, lines)
+    return lines
+
+
+def PullRef(pull):
+    return '+refs/pull/%d/merge' % pull
+
+
+def Repo(repo):
+    return 'https://github.com/%s' % repo
+
+
+def Checkout(repo, branch, pull):
+    if bool(branch) == bool(pull):
+        raise ValueError('Must specify one of --branch or --pull')
+    if pull:
+        ref = PullRef(pull)
+    else:
+        ref = branch
+
+    git = 'git'
+    Subprocess([git, 'init', repo])
```

Comment:
actually, my example is wrong. needs to be either `go/src/github.com` + repo or `go/src/k8s.io`